src/ConstructionBase.jl

```diff
@@ -24,129 +24,145 @@ for (name, path) in [
     end
 end
 
-@generated function constructorof(::Type{T}) where T
-    getfield(parentmodule(T), nameof(T))
+@static if isdefined(Base, Symbol("@assume_effects"))
+    using Base: @assume_effects
+else
+    macro assume_effects(_, ex)
+        Base.@pure ex
+    end
 end
+struct PropertyNames{syms} end
 
-constructorof(::Type{<:Tuple}) = tuple
-constructorof(::Type{<:NamedTuple{names}}) where names =
-    NamedTupleConstructor{names}()
+struct PropertyIndices{inds} end
 
-struct NamedTupleConstructor{names} end
+const PropertyKeys{K} = Union{PropertyNames{K},PropertyIndices{K}}
+
+PropertyNames(::Type{T}) where {T} = PropertyNames{fieldnames(T)}()
+PropertyNames(@nospecialize(T::Type{<:NamedTuple})) = PropertyNames{T.parameters[1]}()
```

Comment:
Why `T.parameters[1]` instead of fieldnames?

Reply:
Under the hood this is what `fieldnames` does on `DataType`. We're skipping some steps to get exactly what we want

Reply:
Ok got it thanks!